src/shred_collector/shred_receiver.zig

```diff
@@ -0,0 +1,202 @@
+const std = @import("std");
+const network = @import("zig-network");
+const sig = @import("../lib.zig");
+
+const bincode = sig.bincode;
+const layout = sig.shred_collector.shred_layout;
+
+const Allocator = std.mem.Allocator;
+const ArrayList = std.ArrayList;
+const Atomic = std.atomic.Value;
+const KeyPair = std.crypto.sign.Ed25519.KeyPair;
+const Socket = network.Socket;
+
+const Channel = sig.sync.Channel;
+const Logger = sig.trace.Logger;
+const Packet = sig.net.Packet;
+const Ping = sig.gossip.Ping;
+const Pong = sig.gossip.Pong;
+const RepairMessage = sig.shred_collector.RepairMessage;
+const Slot = sig.core.Slot;
+const SocketThread = sig.net.SocketThread;
+
+const endpointToString = sig.net.endpointToString;
+
+/// Analogous to [ShredFetchStage](https://github.com/anza-xyz/agave/blob/aa2f078836434965e1a5a03af7f95c6640fe6e1e/core/src/shred_fetch_stage.rs#L34)
+pub const ShredReceiver = struct {
+    allocator: Allocator,
+    keypair: *const KeyPair,
+    exit: *Atomic(bool),
+    logger: Logger,
+    repair_socket: Socket,
+    tvu_socket: Socket,
+    outgoing_shred_channel: *Channel(ArrayList(Packet)),
+    shred_version: *const Atomic(u16),
+
+    const Self = @This();
+
+    /// Run threads to listen/send over socket and handle all incoming packets.
+    /// Returns when exit is set to true.
+    pub fn run(self: *Self) !void {
+        defer self.logger.err("exiting shred receiver");
+        errdefer self.logger.err("error in shred receiver");
+
+        var sender = try SocketThread
+            .initSender(self.allocator, self.logger, self.repair_socket, self.exit);
+        defer sender.deinit();
+        var repair_receiver = try SocketThread
+            .initReceiver(self.allocator, self.logger, self.repair_socket, self.exit);
+        defer repair_receiver.deinit();
+
+        const num_tvu_receivers = 2;
```

Comment:
2bf571c6c6c975eb64b916d9b640b5ee40ac9b34

I moved it to the top level of the file because I think that's what your main intention was here. But I used lower snake case because that's what the zig style guide calls for. Also I made it private, not pub because I'm not sure why it needs to be exposed.

Reply:
private makes sense - but feel like we should stick to global variables to be full caps since thats where we use everywhere else

Reply:
I don't think we need to adhere perfectly to the zig style guide if we prefer a different approach. But if we're going to deviate from the zig style guide, we should be explicit about it in our style guide.

Typically, I would follow the zig style guide unless there's a good reason not to. Personally I don't see much reason to use all caps, so I was leaning towards gradually moving the codebase towards the standard. But I'm fine with using uppercase.

If we're deviating from the zig style guide, the main thing I'd want to do is stop using camel case for functions. I think snake case is more readable. Anyway, I'm not too worried about it.

Reply:
89759aefac786e84090a923520631eeb6b2f5e06

Reply:
yeah i dont think we need to stick to zig style guide unless it makes sense (were kind of already breaking it by using structs everywhere vs the full file thing they do)

also, i agree snake_case for functions is better - we originally wanted to do it that way - the problem is stdlib is camelCase - so throughout our codebase youll see snake_case() and camelCase() calls which hurts to look at